### PR TITLE
#7: Compile time import for transform

### LINKED ISSUE #7: Compile time import for transform

### LINKED ISSUE #34: Support Spark 3.2.1

### PR DESCRIPTION
**Pull request by benedeki**
* support for Spark 3.2.1
* using `sbt-projectmatrix` for cross-build of the modules
* `HofsAdapter` -> `TransformAdapter`
* Shims for `transform` based on Spark version

Closes #7
Closes #34

Replaces #33
Replaces #35
Replaces #36
Replaces #37

`build.yml` needs to change I suppose.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -50,9 +50,9 @@ spark.listenerManager.register(myListener)
 >Different _Scala_ variants of the _Schema Utils_ are compiled against different _Spark_, _Json4s_ and _Jackson_ versions.
 >2.13 is not yet fully supported:
 >
->| | Scala 2.11 | Scala 2.12 | Scala 2.13 | 
+>| | Scala 2.11 | Scala 2.12 | Scala 2.12 |
```
Comment (benedeki) [inline README.md:53]: I am still not sure the table is correct... 🤔

File: spark-commons/src/main/scala-spark2-jvm/za/co/absa/spark/commons/adapters/TransformAdapter.scala [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+package za.co.absa.spark.commons.adapters
+
+import org.apache.spark.sql.Column
+import za.co.absa.spark.hofs.{transform => hofsTransform}
+
+trait TransformAdapter {
+  /**
+   * For Spark versions prior 3.0.0, delegates to {{{hofs.transform()}}}
+   */
+  def transform: (Column, Column => Column) => Column = hofsTransform
```
Comment (benedeki) [inline spark-commons/src/main/scala-spark2-jvm/za/co/absa/spark/commons/adapters/TransformAdapter.scala:25]: Wondering if to include `SparkVersionGuard` call here. 🤔

Comment (AdrianOlosutean) [reply spark-commons/src/main/scala-spark2-jvm/za/co/absa/spark/commons/adapters/TransformAdapter.scala:26]: Don't think it's needed. I also think the comments are not so necessary

Comment (AdrianOlosutean) [reply spark-commons/src/main/scala-spark2-jvm/za/co/absa/spark/commons/adapters/TransformAdapter.scala:26]: Don't think it's needed. I also think the comments are not so necessary

File: project/Dependencies.scala [hunk 3]
```diff
@@ -17,19 +17,25 @@ import sbt._
...
-    "org.scala-lang"     % "scala-compiler" % scalaVersion,
-    "org.scalatest"      %% "scalatest" % "3.1.0" % Test
+  def sparkDependencies(sparkVersion: String): Seq[ModuleID] = Seq(
+    "org.apache.spark"      %% "spark-core"  % sparkVersion % "provided",
+    "org.apache.spark"      %% "spark-sql"   % sparkVersion % "provided"
   )
+
+  def sparkCommonsDependencies(sparkVersion: String): Seq[ModuleID] = {
+    Seq(
+      "za.co.absa"          %% "spark-hats"  % "0.2.2"
```
Comment (AdrianOlosutean) [inline project/Dependencies.scala:37]: Is it possible to have a condition for the spark-hats to only be included in the spark3 module?

Comment (AdrianOlosutean) [inline project/Dependencies.scala:37]: Is it possible to have a condition for the spark-hats to only be included in the spark3 module?

**Review by AdrianOlosutean — COMMENTED** [review]
Minor things

File: project/Dependencies.scala (hunk 3 above)
Comment (benedeki) [reply project/Dependencies.scala:37]: Technically It is possible to have conditional dependencies. I actually had it like that for a moment, until I found out hats is used in Explode tools too. 😉

File: spark-commons/src/main/scala-spark2-jvm/za/co/absa/spark/commons/adapters/TransformAdapter.scala (hunk 2 above)
Comment (benedeki) [reply spark-commons/src/main/scala-spark2-jvm/za/co/absa/spark/commons/adapters/TransformAdapter.scala:26]: Agreed not needed. At least not for now.

File: build.sbt [hunk 4]
```diff
@@ -17,40 +17,39 @@ ThisBuild / organization := "za.co.absa"
...
     publish / skip := true
   )
 
-lazy val sparkCommons = (project in file("spark-commons"))
-  .settings(
-    name := "spark-commons",
-    libraryDependencies ++= sparkCommonslibraryDependencies(scalaVersion.value),
-    (Compile / compile) := ((Compile / compile) dependsOn printSparkScalaVersion).value // printSparkScalaVersion is run with compile
-  ).dependsOn(sparkCommonsTest)
+lazy val `sparkCommons` = (projectMatrix in file("spark-commons"))
```
Comment (AdrianOlosutean) [inline build.sbt:41]: I would like to suggest if possible to have these vals as `spark-commons` and `spark-commons-test`

**Review by AdrianOlosutean — DISMISSED** [review]

**Review by AdrianOlosutean — APPROVED** [review]
Nice

**Pull request merged**